### PR TITLE
[aptos-cli] Init tool

### PR DESCRIPTION
**Pull request by gregnazario**
## Motivation

Create the first of having a persistent configuration for the CLI.  The first usecase is to save the private key wanting to be used in the CLI.  In the future, we can support vault or more safety friendly ways to store this information in the CLI.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

```
$ aptos init --help
aptos-init 0.1.0
Tool to initialize current directory for the aptos tool

USAGE:
    aptos init

OPTIONS:
    -h, --help       Print help information
    -V, --version    Print version information

$ aptos init
Aptos already initialized, do you want to overwrite the existing config? [yes/no] >
y
Enter your private key as a hex literal (0x...) [No input: Generate new key]
Aptos is now set up!  Run `aptos help` for more information about commands
{
  "Result": "Success"
}

$ cat .aptos/config.yml
---
private_key: "0x0000000000000000000000000000000000000000000000000000000000000151"

$ aptos op key extract-peer --output-file peer_info.yaml
"peer_info.yaml" already exists, are you sure you want to overwrite it? [yes/no] >
yes
Using .aptos/config.yml private key
{
  "Result": {
    "34dbec45e06025bf64be9d0bd288c42c84d4dd9e34cd1dc394801ee616e0d45f": {
      "addresses": [],
      "keys": [
        "0x34dbec45e06025bf64be9d0bd288c42c84d4dd9e34cd1dc394801ee616e0d45f"
      ],
      "role": "Upstream"
    }
  }
}
```

### CODE REVIEW

File: crates/aptos-crypto/src/traits.rs [hunk 1]
```diff
@@ -74,16 +77,36 @@ pub trait ValidCryptoMaterialStringExt: ValidCryptoMaterial {
...
     /// bytes before checking if we can convert.
     fn from_encoded_string(encoded_str: &str) -> std::result::Result<Self, CryptoMaterialError> {
-        let bytes_out = ::hex::decode(encoded_str);
+        let mut start = 0;
+        if encoded_str.starts_with("0x") {
+            start = 2;
+        }
+
+        let hex_len = encoded_str.len() - start;
+        const ENCODED_KEY_SIZE: usize = PRIVATE_KEY_SIZE * 2;
```
Comment (CapCap) [inline crates/aptos-crypto/src/traits.rs:86]: can this be a top level const?

File: crates/aptos-crypto/src/traits.rs [hunk 2]
```diff
@@ -74,16 +77,36 @@ pub trait ValidCryptoMaterialStringExt: ValidCryptoMaterial {
...
+        // If the string is too short, pad it
+        let encoded_str = if hex_len < ENCODED_KEY_SIZE {
+            let mut hex_str = String::with_capacity(ENCODED_KEY_SIZE);
+            for _ in 0..ENCODED_KEY_SIZE - hex_len {
```
Comment (CapCap) [inline crates/aptos-crypto/src/traits.rs:90]: I believe you can pad with 
```
assert_eq!("0000012345", format!("{:0>10}", 12345));
```

File: crates/aptos/src/common/init.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+};
+use aptos_crypto::{x25519::PrivateKey, ValidCryptoMaterialStringExt};
+use clap::Parser;
+
+/// Tool to initialize current directory for the aptos tool
+#[derive(Debug, Parser)]
+pub struct InitTool {}
+
+impl InitTool {
+    pub async fn execute(&self) -> CliResult {
```
Comment (CapCap) [inline crates/aptos/src/common/init.rs:20]: self is never used, these can all be class methods

File: crates/aptos/src/common/init.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            }
+            CliConfig::load()?
+        } else {
+            CliConfig::default()
+        };
+
+        eprintln!("Enter your private key as a hex literal (0x...) [No input: Generate new key]");
+        let input = read_line()?;
+        let input = input.trim();
+        let private_key = if input.is_empty() {
```
Comment (CapCap) [inline crates/aptos/src/common/init.rs:40]: this could be a match

File: crates/aptos-crypto/src/traits.rs (hunk 2 above)
Comment (gregnazario) [reply crates/aptos-crypto/src/traits.rs:90]: Can't cause it's a string, there's a long convoluted problem around trait inheritance here.

File: crates/aptos/src/common/init.rs (hunk 3 above)
Comment (gregnazario) [reply crates/aptos/src/common/init.rs:20]: Will fix with all of the other commands in a future PR.

File: crates/aptos/src/common/init.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                return Ok(());
+            }
+            CliConfig::load()?
+        } else {
+            CliConfig::default()
+        };
+
+        eprintln!("Enter your private key as a hex literal (0x...) [No input: Generate new key]");
+        let input = read_line()?;
+        let input = input.trim();
```
Comment (CapCap) [inline crates/aptos/src/common/init.rs:39]: this can be one line

Comment (gregnazario) [reply crates/aptos/src/common/init.rs:39]: Can't cause it needs to live longer :(

File: crates/aptos/src/common/init.rs (hunk 4 above)
Comment (gregnazario) [reply crates/aptos/src/common/init.rs:40]: I'm not sure it'll be more readable though.  I'll leave it like this for now.

**Review by kent-white — APPROVED** [review]

File: crates/aptos-crypto/src/traits.rs (hunk 2 above)
Comment (davidiw) [reply crates/aptos-crypto/src/traits.rs:92]: who or what is this for? I'm a little worried about this as it is within the crypto crate where we anticipate things are done ... sensibly.

File: crates/aptos-crypto/src/traits.rs [hunk 6]
```diff
@@ -155,20 +180,20 @@ pub fn signing_message<T: CryptoHash + Serialize>(message: &T) -> Vec<u8> {
...
+// This unsightly turbofish type parameter is the precise constraint
+// needed to require that there exists an
+//
+// ```
+// impl From<&MyPrivateKeyMaterial> for MyPublicKeyMaterial
+// ```
+//
+// declaration, for any `MyPrivateKeyMaterial`, `MyPublicKeyMaterial`
+// on which we register (respectively) `PublicKey` and `PrivateKey`
+// implementations.
```
Comment (davidiw) [inline crates/aptos-crypto/src/traits.rs:192]: why the tab changing? this doesn't look right

**Review by davidiw — CHANGES_REQUESTED** [review]

**Comment by davidiw** [comment]
/cancel

unclear about the tweak to the crypto-material

File: crates/aptos-crypto/src/traits.rs (hunk 6 above)
Comment (gregnazario) [reply crates/aptos-crypto/src/traits.rs]: Fixed it back, strange formatter didn't fix it.

**Review by davidiw — APPROVED** [review]
thanks!

**Comment by gregnazario** [comment]
Turns out I had a test in a rust block in a comment.  I created a new task to track these kinds of issues https://github.com/aptos-labs/aptos-core/issues/503

/land

**Pull request closed without merging**